internal/host/static/repository_host_test.go

```diff
@@ -857,6 +934,82 @@ func TestRepository_ListHosts(t *testing.T) {
 	}
 }
 
+func TestRepository_ListHosts_HostSets(t *testing.T) {
+	conn, _ := db.TestSetup(t, "postgres")
+	rw := db.New(conn)
+	wrapper := db.TestWrapper(t)
+	kms := kms.TestKms(t, conn, wrapper)
+	iamRepo := iam.TestRepo(t, conn, wrapper)
+	_, prj := iam.TestScopes(t, iamRepo)
+
+	// testing for full and empty hostset associations
+	catalogs := TestCatalogs(t, conn, prj.PublicId, 3)
+	catalogA, catalogB, catalogC := catalogs[0], catalogs[1], catalogs[2]
+	hostsA := TestHosts(t, conn, catalogA.PublicId, 3)
+	setA := TestSets(t, conn, catalogA.PublicId, 1)[0]
+	TestSetMembers(t, conn, setA.PublicId, hostsA)
+	for _, h := range hostsA {
+		h.SetIds = []string{setA.PublicId}
+	}
+	hostsB := TestHosts(t, conn, catalogB.PublicId, 3)
+
+	// testing for mixed hosts with individual hostsets and empty hostsets
+	hostsC := TestHosts(t, conn, catalogC.PublicId, 5)
+	hostsC0 := TestHosts(t, conn, catalogC.PublicId, 2)
+	setC := TestSets(t, conn, catalogC.PublicId, 5)
+	for i, h := range hostsC {
+		h.SetIds = []string{setC[i].PublicId}
+		TestSetMembers(t, conn, setC[i].PublicId, []*Host{hostsC[i]})
+	}
+	hostsC = append(hostsC, hostsC0...)
+
+	tests := []struct {
+		name      string
+		in        string
+		want      []*Host
+		wantIsErr errors.Code
+	}{
+		{
+			name: "with-hostsets",
+			in:   catalogA.PublicId,
+			want: hostsA,
```

Comment:
Nit, and FYI (This is my opinion and other's opinions may differ.  I'm more than happy to discuss further regarding if this makes following the expected in and out of each test easier or harder.):

When the test setup is simply creating variables for a number of inputs to pair with a number of precomputed outputs then much of the benefit of table driven tests is removed since you could have easily just written out 3 different test cases with potentially fewer variable names and inlined the test logic which tends to be a little easier to follow.


Much of the benefit of the table driven tests is to be able to easily see: "For this input, I expect that output." which makes it clearer when a test case might not be covered.

For example, if the tests struct had `getWantedHosts   func(*HostCatalog) []*Host` then you could have a test cases that looked like:

```
{
	name: "with-hostsets",
	getWantedHosts:   func(c *HostCatalog) []*Host {
           hosts := TestHosts(t, conn, c.PublicId, 3)
	   s := TestSets(t, conn, c.PublicId, 1)[0]
	   TestSetMembers(t, conn, s.PublicId, hosts)
	   for _, h := range hosts {
		  h.SetIds = []string{s.PublicId}
	   }
           return hosts
         },
}
```

and in t.Run you could have:
```
...
cat := TestCatalogs(t, conn, prj.PublicId, 1)[0]
want := tt.setupWantedHosts(cat)
...
```

While the getWantedHosts is a bit clunky, it at least it pairs input with the output.  I realize other tests could do better in this regard, especially ones that I've written, so again I'd be super happy to discuss further.

Reply:
Yeah I prefer this method a lot better because it gets rid of a lot of repetitive code. The one nit I have is that the function calls `TestSetMembers` to create the `host_set_member` object as part of its call, which could be confusing in the sense of clarity (function name suggests nothing about actually setting the hostset-host association). I prefer this method over what I currently have because it reduces a lot of repetition.